service/history/api/recordactivitytaskstarted/api.go

```diff
@@ -189,18 +193,32 @@ func recordActivityTaskStarted(
 	}
 
 	if !pollerDeployment.Equal(wfDeployment) {
-		// Task is redirected, see if a transition can start.
-		if err := mutableState.StartDeploymentTransition(pollerDeployment); err != nil {
-			if errors.Is(err, workflow.ErrPinnedWorkflowCannotTransition) {
-				// This must be a task from a time that the workflow was unpinned, but it's
-				// now pinned so can't transition. Matching can drop the task safely.
-				return nil, false, serviceerrors.NewObsoleteMatchingTask(err.Error())
-			}
+		// Task is redirected, see if this activity should start a transition on the workflow. The
+		// workflow transition happens only if the workflow TQ's current deployment is the same as
+		// the poller deployment. Otherwise, it means the activity is independently versioned, we
+		// allow it to start without affecting the workflow.
+		wfTqCurrentDeployment, err := getTaskQueueCurrentDeployment(ctx,
+			request.NamespaceId,
+			mutableState.GetExecutionInfo().GetTaskQueue(),
+			enumspb.TASK_QUEUE_TYPE_WORKFLOW,
+			matchingClient)
+		if err != nil {
+			// Let matching retry
 			return nil, false, err
 		}
-		// This activity started a transition, make sure the MS changes are written but
-		// reject the activity task.
-		return nil, true, nil
+		if pollerDeployment.Equal(wfTqCurrentDeployment) {
```

Comment:
(For later) @dnr there is a race condition here: both TQ's current deployment is updated, but the wf TQ's user data is not yet updated. Ideally, in this case we want to dispatch the activity to the old deployment rather than running it in the new deployment as independent. One way to fix that is to sync wf TQs' user data before activity TQ's user data. Wdyt?

Reply:
If we consider propagation I'm not even sure that's enough. See comment below.